shell/platform/linux/fl_key_embedder_responder.h

```diff
@@ -15,10 +15,20 @@
 
 constexpr int kMaxConvertedKeyData = 3;
 
-typedef std::function<void(const FlutterKeyEvent* event,
-                           FlutterKeyEventCallback callback,
-                           void* user_data)>
-    EmbedderSendKeyEvent;
+// The signature of a function that FlKeyEmbedderResponder calls on every key
+// event.
+//
+// The `send_key_event_user_data` is a pointer managed by the object that
+// creates FlKeyEmbedderResponder and opaque to FlKeyEmbedderResponder.
+//
+// The callback_user_data is a pointer created and managed by
+// FlKeyEmbedderResponder and opaque to the `EmbedderSendKeyEvent` function.
+// After the event is processed, the `callback` should be invoked with
+// `callback_user_data`.
+typedef void (*EmbedderSendKeyEvent)(const FlutterKeyEvent* event,
+                                     FlutterKeyEventCallback callback,
+                                     void* callback_user_data,
+                                     void* send_key_event_user_data);
```

Comment:
Callbacks should really only need one `user_data` baton, but it's `void*` so it can be anything at all; rather than changing the signature and adding a field, change the underlying data that we pass through `user_data` to a struct that contains both fields.

Reply:
The two `user_data` are different. As I've written in the doc, `send_key_event_user_data` is created by the keyboard manager and is opaque to the responder, while `callback_user_data` is created by the responder and opaque to the keyboard manager. Merging them into one struct seems impossible to me. Or, can you tell me how you expect https://github.com/flutter/engine/pull/49293/files#diff-635fefbc86fd31e340d118be9de7b011b099f9b27c43bf06ac23195f50b2adc2R602 to change?

Reply:
I'd change `fl_key_embedder_responder_new` to take a single `void* user_data` and I would pass through a struct like this at the call site, then unpack it in that code:
```objc
struct {
  FLKeyboardManager* keyboard_manager,  // you'd use this as self on line 604.
  void* callback_user_data, // the user_data from the embedder API, used on line 608.
};
```

Reply:
The `callback_user_data` is not passed in at `fl_key_embedder_responder_new`, but generated at every key event. The user data passed in at `fl_key_embedder_responder_new` is `send_key_event_user_data`. It's not possible to group `keyboard_manager` and `callback_user_data` into a single pass-through struct.

(Yeah I understand that it's a pretty entangled structure... maybe trying to refactor based on my current PR would help understanding.)

Reply:
Will ping over chat to discuss. There's probably a reason why it's possible to pass two params to a method but pack them into a struct and pass that, but I'm missing it in the limited context of the diff.

Reply:
Alright; discussed in person and this makes more sense. What I was missing was that there are two separate callbacks here with two separate opaque user_data blobs, one that's effectively an output that the callback is receiving and the other which is an input to the callback being passed in. Thanks for walking me through this.